bpo-43574: Dont overallocate list literals

chadnetzer opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->

Before v3.9, list literals weren't over-allocated, but this behavior regressed starting with v3.9:

Switched to initializing list literals w/ LIST_EXTEND
https://bugs.python.org/issue39320
https://github.com/python/cpython/pull/17984

Commit where over-allocation of list literals first appeared
https://bugs.python.org/issue38328
https://github.com/python/cpython/pull/17114
https://github.com/python/cpython/commit/6dd9b64770af8905bef293c81d541eaaf8d8df52

This changes to the original behavior of not over-allocating lists for list literals, by not over-allocating any 0-length list that is extended.  Technically this is a change in behavior, as lists that are initialized to length 0, and then immediately extended, will not be over-allocated on that initial extension.  This change in behavior is likely innocuous, and possibly desirable, whereas the long-standing Python behavior of not using over-allocation for list literals is almost certainly desired.

Adds a test_overallocation test for lists to find regressions (assuming the regression doesn't alter the behavior of both the list-literal and list-initialization from a known length).

<!-- issue-number: [bpo-43574](https://bugs.python.org/issue43574) -->
https://bugs.python.org/issue43574
<!-- /issue-number -->

chadnetzer committed:
Don't overallocate when extending an empty list

This restores the behavior of Python versions before v3.9, where a list
initialized from a list literal allocates memory for exactly as many
elements as it contains (until a list append or other operation that
changes the list size).

The interpreter was changed in v3.9 to use the LIST_EXTEND bytecode to
initialize a new list from a literal, which caused it to overallocate.
By not performing this over-allocation on an empty list, we restored the
original behavior (though now explicitly creating an empty list and
extending it once will also not result in an over-allocation).

chadnetzer committed:
Add testcase for bpo-43574 list overallocaton

chadnetzer committed:
Skip overallocation calc when not overallocating

pitrou commented on Lib/test/test_list.py:
```diff
@@ -196,6 +196,15 @@ def test_preallocation(self):
+
+        # bpo-43574: Don't overallocate for list literals
+        iterable = [1,2,3]
+        self.assertEqual(sys.getsizeof(iterable), sys.getsizeof(list(iterable)))
```
Is there a reason you test with two different hardcoded literals?

chadnetzer replied:
It's really just a minimal set of test conditions to show this specific regression; the behavior change starts with len-3 literals, as len-2 and below are special-cased to use BUILD_LIST: https://github.com/python/cpython/blob/9a50ef43e42ee32450a81ce13ed5a0729d3b84e8/Python/compile.c#L3721

I'll see if I can generalize those tests a bit more without going overboard; probably to test that initializing a new list from a known-length iterable doesn't over-allocate (ie. catching this if it regresses again for other cases), and also maybe confirming over-allocation when expected.

I'll also take care of requests 1 and 2 (CLA and NEWS item)

pitrou reviewed (changes requested):
The proposed approach looks fine to me. Two things:
1. Can you sign the contributor's agreement?
2. Can you add a NEWS item? See https://devguide.python.org/committing/?highlight=blurb#updating-news-and-what-s-new-in-python

sweeneyde commented on Objects/listobject.c:
```diff
@@ -58,25 +58,28 @@ list_resize(PyListObject *self, Py_ssize_t newsize)
+    if (newsize == 0 || Py_SIZE(self) == 0) {
+        /* Don't overallocate for lists that start empty or are set to empty. */
+        new_allocated = newsize;
+    } else {
```
```suggestion
    }
    else {
```
https://www.python.org/dev/peps/pep-0007/#code-lay-out

sweeneyde commented on Objects/listobject.c:
```diff
@@ -58,25 +58,28 @@ list_resize(PyListObject *self, Py_ssize_t newsize)
+         * than to the old size.
+         */
+        if (newsize - Py_SIZE(self) > (Py_ssize_t)(new_allocated - newsize))
+            new_allocated = ((size_t)newsize + 3) & ~(size_t)3;
```
Per PEP 7: "All new C code requires braces."

chadnetzer committed:
Add more list overallocation regression tests

Ensure that lists initialized from both list-literals and iterables of known
length are initialized without over-allocation, by comparing them with
allocation of lists initialized from an unknown size where over-allocation is
expected.

chadnetzer committed:
Fix list_extend() init list changes to PEP-7 form

chadnetzer committed:
bpo-43574: Add NEWS

chadnetzer committed:
Put common-case first in logical-or list_resize()

Micro-optimization

pitrou reviewed (approved):
+1, thank you

pablogsal commented on Misc/NEWS.d/next/Core and Builtins/2021-03-21-18-51-30.bpo-43574.mteI-I.rst:
```diff
@@ -0,0 +1,2 @@
+Restores previous list memory behavior where lists initialized from literals
```
```suggestion
``list` objects don't overallocate when starting empty or are set to be empty. This effectively restores previous ``list`` memory behavior for lists initialized from literals.
```

chadnetzer replied:
> ``list`

Minor typo - missing backtick in first \`\`list\`\`

chadnetzer replied:
Added a commit to adopt this NEWS description (slightly modified)

chadnetzer committed:
Adjust NEWS entry language based on PR feedback

chadnetzer committed:
Allow empty list append/insert to overallocate

The list_resize() helper is used by single-item list appends and
inserts, which normally would overallocate (to a capacity of 4) on the
first added element, and then again to 8 when the capacity is filled.
But if this over-allocation is postponed on the first added element, the
current expansion formula would then over-allocate to a capacity of 8 on
the second append/insertion.

List-literals of length 1 and 2 are not created with the list_extend()
function (which then calls list_resize()), but instead built directly as
capacity 1 or 2 lists with the BUILD_LIST opcode.  By excluding the
special case for list_resize() of empty lists when newsize is greater
than 1, its allows list append/insert to continue to over-allocate
without skipping capacity 4.

chadnetzer committed:
Update test_overallocation of empty list append()

pitrou commented on Lib/test/test_list.py:
```diff
@@ -196,6 +196,59 @@ def test_preallocation(self):
+        for literal in test_literals:
+            # Ensure that both list literals, and lists made from an iterable
+            # of known size use the same amount of allocation. It will be
+            # verified later that no over-allocation occurs for list literals.
```
This comment says "verified later", but I see no specific test of no-over-allocation of list literals. There is `SizeofTest` in `test_sys.py` but it only tests two literal lengths.
I would imagine this kind of test:
```python
prev_literal = None
for literal in [[] + test_literals]:
    if prev_literal is not None
        # Allocation of literal N uses one more pointer than allocation of literal N-1
        self.assertEqual(sizeof(literal), sizeof(prev_literal) + struct.calcsize('P'))
    prev_literal = literal
```

chadnetzer replied:
Yes, I should have said "verified below", ie. in same test, after the loop.  It's an implicit test, though, since with the current over-allocation strategy, the list capacity can sometimes equal list length, so it's not universally true that each list-literal will have less memory allocated than a normal list of equivalent length.  So it only can test that at least sometimes the list-literals are created with less capacity then a grown list of equivalent length.

I was avoiding a direct test of memory usage, like you propose, just to keep things general.  But I can certainly add such a direct test, which (afaict) is really the only way to test that all the list-literals are being created without over-allocation.

chadnetzer replied:
@pitrou I ran with your idea, and added a direct test of the memory allocated for a list-literal, which also let me remove the confusing wording about verifying "later", since that's no longer the case.  Thanks for the suggestion.

pitrou reviewed (commented):
Thanks for the update. I would welcome a bit more thorough testing.

chadnetzer committed:
Add direct test of list-literal non-overallocation

Confirm that list-literals aren't over-allocated by directly checking
their used memory, which is the combination of the size of an empty list
and the number of pointers needed for a list of exactly that length.

gpshead commented on Lib/test/test_list.py:
```diff
@@ -196,6 +197,63 @@ def test_preallocation(self):
+
+        # bpo-38373: initialized or grown lists are not always over-allocated.
+        # Confirm that over-allocation occurs at least some of the time.
+        self.assertEqual(True, any(x>0 for x in overalloc_amts))
```
add `msg=repr(overalloc_amts)` to aid debugging if this fails.

gpshead commented on Lib/test/test_list.py:
```diff
@@ -196,6 +197,63 @@ def test_preallocation(self):
+            # Direct check that list-literals do not over-allocate, by
+            # calculating the total size of used pointers.
+            total_ptr_size = len(literal) * struct.calcsize('P')
+            self.assertEqual(sizeof(literal), sizeof([]) + total_ptr_size)
```
add a `msg=repr(literal)` arg to all of the asserts in this loop to aid debugging directly from a test failure log.

gpshead commented on Lib/test/test_list.py:
```diff
@@ -196,6 +197,63 @@ def test_preallocation(self):
+                ]
+
+        overalloc_amts = []
+        for literal in test_literals:
```
no action required: meta: this is where I really wish we had an `absltest.paramtereize` or `pytest.mark.parameterize` available in the stdlib.  it'd run all of these as unique tests with appropriate names instead of a loop that prevents others from running upon first failure.

sweeneyde replied:
I think unittest.subTest() displays parameters on the failing subtests and keeps running through each even on failure.

gpshead commented on Lib/test/test_list.py:
```diff
@@ -196,6 +197,63 @@ def test_preallocation(self):
+        # Confirm that over-allocation occurs at least some of the time.
+        self.assertEqual(True, any(x>0 for x in overalloc_amts))
+
+        # Empty lists should overallocate on initial append/insert (unlike
```
Turn this bit into an additional test method.

gpshead commented on Lib/test/test_list.py:
```diff
@@ -196,6 +197,63 @@ def test_preallocation(self):
+        # bpo-43574: Don't overallocate for list-literals
+        sizeof = sys.getsizeof
+
+        # First handle empty list and empty list-literal cases.  Should have no
```
make this set of initial assertions into its own test method.

gpshead commented on Misc/NEWS.d/next/Core and Builtins/2021-03-21-18-51-30.bpo-43574.mteI-I.rst:
```diff
@@ -0,0 +1,3 @@
+``list`` objects don't overallocate when starting empty and then extended, or
+when set to be empty.  This effectively restores previous ``list`` memory
```
quantify what "previous" means.  3.8 and earlier I believe.

gpshead reviewed (changes requested):
Overall good, some test maintainability and debugging suggestions.
